### PR TITLE
lzma: Use non-streamed mode

### PR DESCRIPTION
**Pull request by rjoleary**
Signed-off-by: Ryan O'Leary <ryanoleary@google.com>

### CODE REVIEW

**Comment by rjoleary** [comment]
Travis's "file" command is different from my "file" command.

File: pkg/lzma/lzma.go [hunk 1]
```diff
@@ -7,40 +7,97 @@ package lzma
...
+// backwards compatible with lzma (assume the --format=lzma flag is used). And,
+// on many distros, lzma points to xz. XZ is the new LZMA.
+var Default = LZMA{
+	XZPath: "xz",
+}
 
 // Decode decodes a byte slice of LZMA data.
-func Decode(encodedData []byte) ([]byte, error) {
-	cmd := exec.Command(lzmaDecode)
+func (l LZMA) Decode(encodedData []byte) ([]byte, error) {
```
Comment (rminnich) [inline pkg/lzma/lzma.go]: l LZMA? Why not a pointer?

File: pkg/lzma/lzma.go [hunk 2]
```diff
@@ -7,40 +7,97 @@ package lzma
...
-	return cmd.Output()
+	out, err := cmd.Output()
+	if err == nil {
+		return out, nil
+	}
+
+	// TODO: This is extremely hacky. This reverses the damage we inflicted
+	// on the compressed lzma file in EncodeLevel.
+	copy(encodedData[5:5+8], []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff})
+	cmd = exec.Command(l.XZPath, args...)
```
Comment (hugelgupf) [inline pkg/lzma/lzma.go]: Why xz decompress twice? If there's a good reason (I don't really get it), please add a comment explaining this.

Comment (rjoleary) [reply pkg/lzma/lzma.go]: I think it is a bug in xz. I'm going to look into it more today.

Comment (hugelgupf) [reply pkg/lzma/lzma.go]: wat

**Comment by rjoleary** [comment]
Abandoning in favour of #90

File: pkg/lzma/lzma.go (hunk 2 above)
Comment (rjoleary) [reply pkg/lzma/lzma.go]: Better explanation: When the xz command compresses, it stores the size at the end of the file and 0xFFFFFFFFFFFFFFFF in the header. EDK2's decompressor is primitive and will try to allocate 0xFFFFFFFFFFFFFFFF bytes and fail. So, the Encode function writes the size in the header which works for EDK2's tool. Unfortunately, xz considers an lzma file which has a size in both places corrupt, but will still decompress it and return exit status 1 (false positive). The second decompression overwrites the header to see if first's exit status is a false positive.

We're not going to do this and instead use the Go implementation of LZMA and upstream any issues we encounter.

**Pull request closed without merging**